walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,135 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.*; 
+import java.util.Comparator;
+import java.util.Collections;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    List<TimeRange> mandatoryAttendeeAvailability = new ArrayList<>();
+    List<TimeRange> allAttendeeAvailability = new ArrayList<>();
+
+    //if the meeting is scheduled for longer than a day, it cannot be scheduled
+    if (request.getDuration() > 24*60) {
+      return allAttendeeAvailability;
+    } 
+
+    //if no attendees or no meetings, any time of day works
+    else if ((request.getAttendees().size() == 0 && request.getOptionalAttendees().size() == 0) || events.size() == 0) {
+      allAttendeeAvailability.add(TimeRange.WHOLE_DAY);
+      return allAttendeeAvailability;
+    } 
+
+    //otherwise, find all possible meeting times   
+    else {
+
+      //first find the availability for mandatory attendees only 
+      List<TimeRange> mandatoryAttendeeMeetingTimes = 
+          findAndSortConflictingMeetings(events, request, false);
+      
+      fixNestedOrOverlappingMeetings(mandatoryAttendeeMeetingTimes);
+      mandatoryAttendeeAvailability = findAvailabilityBetweenMeetings(mandatoryAttendeeMeetingTimes, request);
+
+      //if optional attendees, find availability for all attendees and compare to availability for mandatory attendees 
+      if (request.getOptionalAttendees().size() > 0) {
+        List<TimeRange> allAttendeeMeetingTimes = 
+            findAndSortConflictingMeetings(events, request, true);
+
+        fixNestedOrOverlappingMeetings(allAttendeeMeetingTimes);
+        allAttendeeAvailability = findAvailabilityBetweenMeetings(allAttendeeMeetingTimes, request);
+
+        //compare mandatory attendees' availability with optional attendees' availability
+        if (allAttendeeAvailability.size() > 0 || request.getAttendees().size() == 0) {
+          return allAttendeeAvailability;
+        }
+      }
+    }
+    return mandatoryAttendeeAvailability;
+  }
+
+  //only keep meetings that have overlapping attendees, and sort by start time 
+  public List<TimeRange> findAndSortConflictingMeetings(Collection<Event> events, MeetingRequest request, Boolean allAttendees) {
```

Comment:
Since these helper functions are only used inside this class, it's best to limit their visibility to private instead of making it public.